### PR TITLE
HPCC-8983 Use visibility flags in gcc

### PR DESCRIPTION
**Pull request by ghalliday**

### CODE REVIEW

**Comment by ghalliday** [comment]
@richardkchapman for merging into 6.4

**Comment by ghalliday** [comment]
See also https://github.com/hpcc-systems/LN/pull/811

**Comment by ghalliday** [comment]
@ghalliday need to check anything that can be thrown is exported, and run the full regression suite with this enabled.

**Comment by ghalliday** [comment]
note, the problem with thrown exceptions is currently causing problems building the system in some cases - so definitely needs looking at.

File: roxie/roxiemem/CMakeLists.txt [hunk 1]
```diff
@@ -40,6 +40,10 @@ include_directories (
          ./../../testing/unittests
     )
 
+if (CMAKE_COMPILER_IS_GNUCC)
+  ADD_DEFINITIONS( -fvisibility=hidden)
```
Comment (richardkchapman) [inline roxie/roxiemem/CMakeLists.txt]: Why is this library different?

File: roxie/roxiemem/roxiemem.cpp [hunk 2]
```diff
@@ -1265,7 +1265,7 @@ struct HeapletStats
 
 class CHeap;
 static void noteEmptyPage(CHeap * heap);
-class Heaplet : public HeapletBase
+class __attribute__ ((visibility ("hidden"))) Heaplet : public HeapletBase
```
Comment (richardkchapman) [inline roxie/roxiemem/roxiemem.cpp]: (and if they are, should we use DECL_LOCAL ?)

File: system/include/platform.h [hunk 3]
```diff
@@ -508,4 +508,22 @@ typedef __int64 cycle_t;
 #endif
 #endif
 
+//Use DECL_EXCEPTION for objects which are thrown if they are not exported from a dll
```
Comment (richardkchapman) [inline system/include/platform.h]: Could you expand on this a bit... also when to use DECL_LOCAL ?

File: system/jlib/CMakeLists.txt [hunk 4]
```diff
@@ -197,6 +197,10 @@ if ( ${HAVE_LIBDL} )
 target_link_libraries ( jlib dl)
 endif ( ${HAVE_LIBDL} )
 
+if (CMAKE_COMPILER_IS_GNUCC)
+  ADD_DEFINITIONS( -fvisibility=hidden)
```
Comment (richardkchapman) [inline system/jlib/CMakeLists.txt]: is this needed? Already set globally....

**Comment by richardkchapman** [comment]
Is there / do we need a corresponding change in the LN repo?

**Comment by richardkchapman** [comment]
@ghalliday Some questions

File: system/include/platform.h [hunk 5]
```diff
@@ -508,4 +508,22 @@ typedef __int64 cycle_t;
...
+#ifdef _MSC_VER
+ #define DECL_EXPORT __declspec(dllexport)
+ #define DECL_IMPORT __declspec(dllimport)
+ #define DECL_LOCAL
+ #define DECL_EXCEPTION
+#elif __GNUC__ >= 4
```
Comment (richardkchapman) [inline system/include/platform.h]: Should we also make the use of -fvisibility=hidden in the makefile dependent on gnuc v4 ?

**Comment by richardkchapman** [comment]
The rebase seems to have a commit in it that does not belong

File: roxie/roxiemem/CMakeLists.txt (hunk 1 above)
Comment (ghalliday) [inline roxie/roxiemem/CMakeLists.txt]: I suspect that is left in from some testing.  I'll remove

File: system/jlib/CMakeLists.txt (hunk 4 above)
Comment (ghalliday) [inline system/jlib/CMakeLists.txt]: It could now go - it has been set for a long time.  I will clean up.

File: system/include/platform.h [hunk 6]
```diff
@@ -508,7 +508,11 @@ typedef __int64 cycle_t;
 #endif
 #endif
 
-//Use DECL_EXCEPTION for objects which are thrown if they are not exported from a dll
+//Use the following DECL_XXX to annotate items that are exported from dlls, or objects that are thrown:
+//DECL_EXPORT for items that are exported from the dll containing the current file being compiled
+//DECL_EXPORT for items that are imported from another dll
```
Comment (richardkchapman) [inline system/include/platform.h]: DECL_IMPORT you mean I think

**Pull request merged**